### PR TITLE
20220809-quic-fixes

### PR DESCRIPTION
**Pull request by douzzer**
fixes for issues introduced in #5384:

added numerous missing `_SMALL_STACK` code paths (PK objects on the stack);

in settings.h, enable `WOLFSSL_SMALL_STACK_STATIC` by default when `WOLFSSL_SMALL_STACK` is defined (`NO_WOLFSSL_SMALL_STACK_STATIC` to override);

fixes for unsafe `strcat()`s in tests/quic.c;

fix for unsafe macro `WOLFSSL_IS_QUIC()`;

fix to exclude quic from `enable-all` when `enable-linuxkm` (quic needs opensslextra, and opensslextra currently only works in-kernel in cryptonly builds);

fix for signed/unsigned clash in `wolfSSL_quic_receive()`.

tested with `wolfssl-multi-test.sh ... super-quick-check`

### CODE REVIEW

File: src/internal.c [hunk 1]
```diff
@@ -25903,33 +25903,48 @@ static int HashSkeData(WOLFSSL* ssl, enum wc_HashType hashType,
...
+#endif
 
-                InitDecodedCert(&cert, input + *inOutIdx, dnSz, ssl->heap);
+                InitDecodedCert(cert, input + *inOutIdx, dnSz, ssl->heap);
 
-                if ((ret = GetName(&cert, SUBJECT, dnSz)) != 0) {
-                    FreeDecodedCert(&cert);
-                    return ret;
-                }
+                do {
```
Comment (dgarske) [inline src/internal.c:25920]: Not a fan of the `do {} while(0)` shortcut for break. Can you just use the `if (ret == 0) {} if (ret == 0) {}` flow?

File: src/ocsp.c [hunk 2]
```diff
@@ -616,50 +623,60 @@ WOLFSSL_OCSP_CERTID* wolfSSL_OCSP_cert_to_id(
...
+#endif
 
-    return certId;
+    if (ret != 0) {
+        if (certId != NULL)
+            XFREE(certId, NULL, DYNAMIC_TYPE_OPENSSL);
+        if (certStatus)
+            XFREE(certStatus, NULL, DYNAMIC_TYPE_OPENSSL);
+        return NULL;
+    } else
```
Comment (dgarske) [inline src/ocsp.c:678]: Instead of ` else return certId` just end with `return certId;` and remove else.

File: src/ocsp.c [hunk 3]
```diff
@@ -616,50 +623,60 @@ WOLFSSL_OCSP_CERTID* wolfSSL_OCSP_cert_to_id(
                 issuer->derCert->length);
         ret = AddCA(cm, &derCert, WOLFSSL_USER_CA, 1);
         if (ret != WOLFSSL_SUCCESS) {
-            wolfSSL_CertManagerFree(cm);
-            return NULL;
+            goto out;
         }
     }
 
+    ret = -1;
```
Comment (dgarske) [inline src/ocsp.c:630]: Use `cm->heap` hint for the XMALLOC's and XFREE's below.

File: src/pk.c [hunk 4]
```diff
@@ -6789,7 +6802,7 @@ WOLFSSL_BIGNUM* wolfSSL_DH_6144_prime(WOLFSSL_BIGNUM* bn)
  */
 WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
 {
-    const char prm[] = {
+    static const char prm[] = {
```
Comment (dgarske) [inline src/pk.c:6805]: Why add `static`?

File: src/x509.c [hunk 5]
```diff
@@ -115,14 +115,18 @@ unsigned int wolfSSL_X509_get_extended_key_usage(WOLFSSL_X509* x509)
 /* Returns the number of X509V3 extensions in X509 object, or 0 on failure */
 int wolfSSL_X509_get_ext_count(const WOLFSSL_X509* passedCert)
 {
-    int extCount = 0;
+    int extCount = WOLFSSL_FAILURE;
```
Comment (dgarske) [inline src/x509.c:118]: Using `0` as the init here seemed better. This is a count...

File: src/x509.c [hunk 6]
```diff
@@ -11599,15 +11629,30 @@ int wolfSSL_X509_check_host(WOLFSSL_X509 *x, const char *chk, size_t chklen,
         return WOLFSSL_FAILURE;
     }
 
-    InitDecodedCert(&dCert, x->derCert->buffer, x->derCert->length, NULL);
-    ret = ParseCertRelative(&dCert, CERT_TYPE, 0, NULL);
+#ifdef WOLFSSL_SMALL_STACK
+    dCert = (DecodedCert *)XMALLOC(sizeof(*dCert), x->heap,
```
Comment (dgarske) [inline src/x509.c:11633]: Allocation of DecodedCert should use `DYNAMIC_TYPE_DCERT`

File: src/x509.c [hunk 7]
```diff
@@ -1723,26 +1742,33 @@ int wolfSSL_X509_get_ext_by_NID(const WOLFSSL_X509* x509, int nid, int lastPos)
         return WOLFSSL_FATAL_ERROR;
     }
 
-    InitDecodedCert( &cert, rawCert, (word32)outSz, 0);
+#ifdef WOLFSSL_SMALL_STACK
+    cert = (DecodedCert *)XMALLOC(sizeof(*cert), x509->heap,
```
Comment (dgarske) [inline src/x509.c:1746]: Allocation of DecodedCert should use `DYNAMIC_TYPE_DCERT`

File: src/x509.c [hunk 8]
```diff
@@ -11632,24 +11681,40 @@ int wolfSSL_X509_check_ip_asc(WOLFSSL_X509 *x, const char *ipasc,
         ret = WOLFSSL_SUCCESS;
     }
 
+#ifdef WOLFSSL_SMALL_STACK
+    if (ret == WOLFSSL_SUCCESS) {
+        dCert = (DecodedCert *)XMALLOC(sizeof(*dCert), x->heap,
```
Comment (dgarske) [inline src/x509.c:11686]: Allocation of DecodedCert should use `DYNAMIC_TYPE_DCERT`

File: tests/quic.c [hunk 9]
```diff
@@ -951,15 +951,15 @@ static int QuicConversation_step(QuicConversation *conv)
         conv->started = 1;
     }
     if (conv->server->output.len > 0) {
-        QuicTestContext_forward(conv->server, conv->client, conv->rec_log);
+        QuicTestContext_forward(conv->server, conv->client, conv->rec_log, sizeof conv->rec_log);
```
Comment (dgarske) [inline tests/quic.c:954]: `sizeof()` please.

File: tests/quic.c [hunk 10]
```diff
@@ -951,15 +951,15 @@ static int QuicConversation_step(QuicConversation *conv)
...
         n = wolfSSL_quic_read_write(conv->client->ssl);
         if (n != WOLFSSL_SUCCESS) {
             AssertIntEQ(wolfSSL_get_error(conv->client->ssl, 0), SSL_ERROR_WANT_READ);
         }
         return 1;
     }
     else if (conv->client->output.len > 0) {
-        QuicTestContext_forward(conv->client, conv->server, conv->rec_log);
+        QuicTestContext_forward(conv->client, conv->server, conv->rec_log, sizeof conv->rec_log);
```
Comment (dgarske) [inline tests/quic.c:962]: `sizeof()` please.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/pk.c (hunk 4 above)
Comment (douzzer) [reply src/pk.c:6805]: the incumbent code was frivolously copying the `const` object to the stack before copying that copy to the caller-supplied `bn` using `wolfSSL_BN_hex2bn()`.  the copy to the stack triggered `gcc -Wframe-larger-than=2048`.

in round#1 I only fixed this one instance, but now that you remind me of it, I've added the `static` attribute to the rest of the `wolfSSL_DH_*_prime()` accessors, by the same rationale.

File: src/internal.c (hunk 1 above)
Comment (douzzer) [reply src/internal.c:25920]: done

File: src/ocsp.c (hunk 2 above)
Comment (douzzer) [reply src/ocsp.c:678]: done

File: src/ocsp.c (hunk 3 above)
Comment (douzzer) [reply src/ocsp.c:630]: done

File: src/x509.c (hunk 5 above)
Comment (douzzer) [reply src/x509.c:118]: done

File: src/x509.c (hunk 6 above)
Comment (douzzer) [reply src/x509.c:11633]: done

File: src/x509.c (hunk 7 above)
Comment (douzzer) [reply src/x509.c:1746]: done

File: src/x509.c (hunk 8 above)
Comment (douzzer) [reply src/x509.c:11686]: done

File: tests/quic.c (hunk 9 above)
Comment (douzzer) [reply tests/quic.c:954]: done

File: tests/quic.c (hunk 10 above)
Comment (douzzer) [reply tests/quic.c:962]: done

**Review by dgarske — APPROVED** [review]

**Pull request merged**